File: 1559-cherry-pickup-ii/cherry-pickup-ii.py
import sys
class Solution:
      
    def cherryPickup(self, grid: List[List[int]]) -> int:
     row=len(grid)
     col=len(grid[0])
     dp=[[[-1 for i in range(col+1)]for j in range(col+1)]for k in range(row+1)]
     for j1 in range(col):
      for j2 in range(col):
       if j1==j2:
        dp[row-1][j1][j2]=grid[row-1][j2]
       else:
        dp[row-1][j1][j2]=grid[row-1][j1]+grid[row-1][j2]
     for i1 in range(row-2,-1,-1):
      for j1 in range(col-1,-1,-1):
       for j2 in range(col):
         maxi=-sys.maxsize
         for x in range(-1,2):
          for y in range(-1,2):
           if j1+x<0 or j1+x>col or j2+y<0 or j2+y>col:
            value=-sys.maxsize
           else:
            value=dp[i1+1][j1+x][j2+y]
           if j1==j2:
            maxi=max(maxi,value+grid[i1][j1])
           else:
            maxi=max(maxi,value+grid[i1][j1]+grid[i1][j2])
           dp[i1][j1][j2]=maxi
     return dp[0][0][col-1]

File: 1559-cherry-pickup-ii/cherry-pickup-ii.cpp
class Solution {
public:
    int solve(int c1,int c2,int r,vector<vector<int>>& grid,vector<vector<vector<int>>>&dp){
        if(c1<0 || c2<0 || c1==grid[0].size() || c2==grid[0].size() || r==grid.size()){ return 0; }
        if(dp[c1][c2][r] !=-1){
            return dp[c1][c2][r];
        }
        int cherries=0;
        if(c1==c2){
            cherries+=grid[r][c1];
        } else {
            cherries+=grid[r][c1]+grid[r][c2];
        }
        int chr1=grid[r][c1];
        int chr2=grid[r][c2];
        grid[r][c1]=0;
        grid[r][c2]=0;

        int ans=0;
        for(int i=-1;i<=1;i++){
            for(int j=-1;j<=1;j++){
                ans=max(ans , solve(c1+i,c2+j,r+1,grid,dp));
            }
        }

        grid[r][c1]=chr1;
        grid[r][c2]=chr2;
        return dp[c1][c2][r]=cherries+ans;
        
    }
    int cherryPickup(vector<vector<int>>& grid) {
        int c=grid[0].size();
        int r=grid.size();
        vector<vector<vector<int>>>dp(c,vector<vector<int>>(c,vector<int>(r,-1)));
        return solve(0,c-1,0,grid,dp);
    }
};